features/main/input_output.feature

```diff
@@ -0,0 +1,212 @@
+    And the JSON should be a superset of:
+    """
+    {
+      "@context": {
+        "@vocab": "http://example.com/docs.jsonld#",
+        "hydra": "http://www.w3.org/ns/hydra/core#",
+        "foo": {
+          "@type": "@id"
+        },
+        "bar": {
+          "@type": "@id"
+        }
+      },
+      "@type": "CustomOutputDto",
+      "foo": "test",
+      "bar": 1
+    }
+    """
+
+  @createSchema
+  Scenario: Get a collection with a custom output
+    Given there are 2 DummyCustomDto
+    When I send a "GET" request to "/dummy_dto_custom_output"
+    Then the response status code should be 200
+    And the response should be in JSON
+    And the header "Content-Type" should be equal to "application/ld+json; charset=utf-8"
+    And the JSON should be a superset of:
+    """
+    {
+      "@context": "/contexts/DummyDtoCustom",
+      "@id": "/dummy_dto_customs",
+      "@type": "hydra:Collection",
+      "hydra:member": [
+        {
+          "foo": "test",
+          "bar": 1
+        },
+        {
+          "foo": "test",
+          "bar": 2
+        }
+      ],
+      "hydra:totalItems": 2
+    }
+    """
+
+  @createSchema
+  Scenario: Create a DummyCustomDto object without output
+    When I add "Content-Type" header equal to "application/ld+json"
+    And I send a "POST" request to "/dummy_dto_custom_post_without_output" with body:
+    """
+    {
+      "lorem": "test",
+      "ipsum": "1"
+    }
+    """
+    Then the response status code should be 201
+    And the response should be empty
+
+  @createSchema
+  Scenario: Create and update a DummyInputOutput
+    When I add "Content-Type" header equal to "application/ld+json"
+    And I send a "POST" request to "/dummy_dto_input_outputs" with body:
+    """
+    {
+      "foo": "test",
+      "bar": 1
+    }
+    """
+    Then the response status code should be 201
+    And the JSON should be a superset of:
+    """
+    {
+      "@context": {
+        "@vocab": "http://example.com/docs.jsonld#",
+        "hydra": "http://www.w3.org/ns/hydra/core#",
+        "baz": {
+          "@type": "@id"
+        },
+        "bat": {
+          "@type": "@id"
+        }
+      },
+      "@type": "OutputDto",
+      "baz": 1,
+      "bat": "test"
+    }
+    """
+    Then I add "Content-Type" header equal to "application/ld+json"
+    And I send a "PUT" request to "/dummy_dto_input_outputs/1" with body:
+    """
+    {
+      "foo": "test",
+      "bar": 2
+    }
+    """
+    Then the response status code should be 200
+    And the JSON should be a superset of:
+    """
+    {
+      "@context": {
+        "@vocab": "http:\/\/example.com\/docs.jsonld#",
+        "hydra": "http:\/\/www.w3.org\/ns\/hydra\/core#",
+        "baz": {
+          "@type": "@id"
+        },
+        "bat": {
+          "@type": "@id"
+        }
+      },
+      "@type": "OutputDto",
+      "baz": 2,
+      "bat": "test"
+    }
+    """
+
+  @createSchema
+  @!mongodb
```

Comment:
Why? FOSUser should be working with MongoDB.

Reply:
Because it's hard to make this test work on both. Small example, I declare a "fake" relation to `Dummy` in `RecoverPasswordOutput` with a doc block:

```
    /**
     * @var \ApiPlatform\Core\Tests\Fixtures\TestBundle\Entity\Dummy
     */
```

I can't specify `Dummy` and `DummyDocument` here, the property metadata factory will take the first type available. This means that the result will not be normalized the same way. 

If I wanted to make this test work on mongodb I have to duplicate everything, there's no point because the only thing that changes is the data provider and yes it definitely works.